File: javascript-sdk/src/tactics/vector.ts
import { VectorStore } from "langchain/vectorstores/base";
import { RebuffError } from "../interface";
import Tactic, { TacticResult } from "./tactic";


export default class Vector implements Tactic {
  name = "vector";

  private vectorStore: VectorStore;
  private similarityThreshold: number;

  constructor(vectorStore: VectorStore, similarityThreshold: number) {
    this.vectorStore = vectorStore;
    this.similarityThreshold = similarityThreshold;
  }

  async execute(input: string): Promise<TacticResult> {
    try {
      const topK = 20;
      const results = await this.vectorStore.similaritySearchWithScore(input, topK);
  
      let topScore = 0;
      let countOverMaxVectorScore = 0;
  
      for (const [_, score] of results) {
        if (score == undefined) {
          continue;
        }
  
        if (score > topScore) {
          topScore = score;
        }
  
        if (score >= this.similarityThreshold && score > topScore) {
          countOverMaxVectorScore++;
        }
      }
  
      return { score: topScore, extraFields: { countOverMaxVectorScore } };
    } catch (error) {
      throw new RebuffError(`Error in getting score for vector tactic: ${error}`);
    }
  }
}

File: python-sdk/rebuff/tactics/vector.py
from typing import Optional

import pinecone
from langchain_community.vectorstores import Pinecone
from langchain_openai import OpenAIEmbeddings

from ..rebuff import TacticName
from .tactic import Tactic, TacticExecution


class Vector(Tactic):
    name = TacticName.VECTOR_DB

    def __init__(self, threshold: float, vector_store: Pinecone):
        self.default_threshold = threshold
        self.vector_store = vector_store

    def execute(self, input: str, threshold_override: float) -> TacticExecution:
        threshold = threshold_override or self.default_threshold
        top_k = 20
        results = self.vector_store.similarity_search_with_score(input, top_k)

        top_score = 0
        count_over_max_vector_score = 0

        for _, score in results:
            if score is None:
                continue

            if score > top_score:
                top_score = score

            if score >= threshold and score > top_score:
                count_over_max_vector_score += 1

        additional_fields = {"countOverMaxVectorScore": count_over_max_vector_score}
        return TacticExecution(score=top_score, additional_fields=additional_fields)


def init_pinecone(api_key: str, index: str, openai_api_key: str) -> Pinecone:
    """
    Initializes connection with the Pinecone vector database using existing (rebuff) index.

    Args:
        api_key (str): Pinecone API key
        index (str): Pinecone index name
        openai_api_key (str): Open AI API key

    Returns:
        vector_store (Pinecone)

    """
    if not api_key:
        raise ValueError("Pinecone apikey definition missing")

    pinecone.Pinecone(api_key=api_key)

    openai_embeddings = OpenAIEmbeddings(
        openai_api_key=openai_api_key, model="text-embedding-ada-002"
    )

    vector_store = Pinecone.from_existing_index(
        index, openai_embeddings, text_key="input"
    )

    return vector_store
